internal/elasticsearch/index/index/models.go

```diff
@@ -0,0 +1,642 @@
+	AnalysisFilter                     jsontypes.Normalized `tfsdk:"analysis_filter"`
+	AnalysisNormalizer                 jsontypes.Normalized `tfsdk:"analysis_normalizer"`
+	Alias                              types.Set            `tfsdk:"alias"`
+	Mappings                           jsontypes.Normalized `tfsdk:"mappings"`
+	SettingsRaw                        jsontypes.Normalized `tfsdk:"settings_raw"`
+	DeletionProtection                 types.Bool           `tfsdk:"deletion_protection"`
+	IncludeTypeName                    types.Bool           `tfsdk:"include_type_name"`
+	WaitForActiveShards                types.String         `tfsdk:"wait_for_active_shards"`
+	MasterTimeout                      customtypes.Duration `tfsdk:"master_timeout"`
+	Timeout                            customtypes.Duration `tfsdk:"timeout"`
+	Settings                           types.List           `tfsdk:"settings"`
+}
+
+type aliasTfModel struct {
+	Name          types.String         `tfsdk:"name"`
+	Filter        jsontypes.Normalized `tfsdk:"filter"`
+	IndexRouting  types.String         `tfsdk:"index_routing"`
+	IsHidden      types.Bool           `tfsdk:"is_hidden"`
+	IsWriteIndex  types.Bool           `tfsdk:"is_write_index"`
+	Routing       types.String         `tfsdk:"routing"`
+	SearchRouting types.String         `tfsdk:"search_routing"`
+}
+
+type settingsTfSet struct {
+	Setting types.Set `tfsdk:"setting"`
+}
+
+type settingTfModel struct {
+	Name  types.String `tfsdk:"name"`
+	Value types.String `tfsdk:"value"`
+}
+
+func (model *tfModel) populateFromAPI(ctx context.Context, indexName string, apiModel models.Index) diag.Diagnostics {
+	model.Name = types.StringValue(indexName)
+	modelAliases, diags := aliasesFromAPI(ctx, apiModel)
+	if diags.HasError() {
+		return diags
+	}
+
+	model.Alias = modelAliases
+
+	if apiModel.Mappings != nil {
+		mappingBytes, err := json.Marshal(apiModel.Mappings)
+		if err != nil {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic("failed to marshal index mappings", err.Error()),
+			}
+		}
+
+		model.Mappings = jsontypes.NewNormalizedValue(string(mappingBytes))
+	}
+
+	diags = setSettingsFromAPI(ctx, model, apiModel)
+	if diags.HasError() {
+		return diags
+	}
+
+	return nil
+}
+
+func aliasesFromAPI(ctx context.Context, apiModel models.Index) (basetypes.SetValue, diag.Diagnostics) {
+	aliases := []aliasTfModel{}
+	for name, alias := range apiModel.Aliases {
+		tfAlias, diags := newAliasModelFromAPI(name, alias)
+		if diags.HasError() {
+			return basetypes.SetValue{}, diags
+		}
+
+		aliases = append(aliases, tfAlias)
+	}
+
+	modelAliases, diags := types.SetValueFrom(ctx, aliasElementType(), aliases)
+	if diags.HasError() {
+		return basetypes.SetValue{}, diags
+	}
+
+	return modelAliases, nil
+}
+
+func setSettingsFromAPI(ctx context.Context, model *tfModel, apiModel models.Index) diag.Diagnostics {
+	modelType := reflect.TypeOf(*model)
+
+	for _, key := range dynamicSettingsKeys {
```

Comment:
I didn't, but probably because I was keeping this analogous to the tf -> api conversion logic. It's an interesting point though.

I had a quick look at this, and I'm concerned that there's a bunch of inconsistencies around how ES handles settings which make this tricky. For example, `number_of_shards` is defined in the schema as an int64, we pass it to ES as an int, but it gets returned as a string which then breaks decoding (since it's an int in the schema). 

I'm sure there's some logic that we could build to make this stuff work, but I'm not sure it will be much nicer than whats in this PR. I can dig in here more if you think it's worthwhile, but I'm not convinced it's worth the time.

Reply:
I trust your judgement. Let's keep it as is.